node/core/approval-voting/src/criteria.rs

```diff
@@ -88,25 +91,105 @@ impl From<OurAssignment> for crate::approval_db::v1::OurAssignment {
 	}
 }
 
-fn relay_vrf_modulo_transcript(relay_vrf_story: RelayVRFStory, sample: u32) -> Transcript {
-	// combine the relay VRF story with a sample number.
-	let mut t = Transcript::new(approval_types::RELAY_VRF_MODULO_CONTEXT);
-	t.append_message(b"RC-VRF", &relay_vrf_story.0);
-	sample.using_encoded(|s| t.append_message(b"sample", s));
+// Combines the relay VRF story with a sample number if any.
+fn relay_vrf_modulo_transcript_inner(
+	mut transcript: Transcript,
+	relay_vrf_story: RelayVRFStory,
+	sample: Option<u32>,
+) -> Transcript {
+	transcript.append_message(b"RC-VRF", &relay_vrf_story.0);
 
-	t
+	if let Some(sample) = sample {
+		sample.using_encoded(|s| transcript.append_message(b"sample", s));
+	}
+
+	transcript
+}
+
+fn relay_vrf_modulo_transcript_v1(relay_vrf_story: RelayVRFStory, sample: u32) -> Transcript {
+	relay_vrf_modulo_transcript_inner(
+		Transcript::new(approval_types::v1::RELAY_VRF_MODULO_CONTEXT),
+		relay_vrf_story,
+		Some(sample),
+	)
+}
+
+fn relay_vrf_modulo_transcript_v2(relay_vrf_story: RelayVRFStory) -> Transcript {
+	relay_vrf_modulo_transcript_inner(
+		Transcript::new(approval_types::v2::RELAY_VRF_MODULO_CONTEXT),
+		relay_vrf_story,
+		None,
+	)
+}
+
+/// A hard upper bound on num_cores * target_checkers / num_validators
+const MAX_MODULO_SAMPLES: usize = 40;
+
+use std::convert::AsMut;
+
+fn clone_into_array<A, T>(slice: &[T]) -> A
+where
+	A: Default + AsMut<[T]>,
+	T: Clone,
+{
+	let mut a = A::default();
+	<A as AsMut<[T]>>::as_mut(&mut a).clone_from_slice(slice);
+	a
+}
+
+struct BigArray(pub [u8; MAX_MODULO_SAMPLES * 4]);
+
+impl Default for BigArray {
+	fn default() -> Self {
+		BigArray([0u8; MAX_MODULO_SAMPLES * 4])
+	}
+}
+
+impl AsMut<[u8]> for BigArray {
+	fn as_mut(&mut self) -> &mut [u8] {
+		self.0.as_mut()
+	}
+}
+
+/// Takes the VRF output as input and returns a Vec of cores the validator is assigned
+/// to as a tranche0 checker.
+fn relay_vrf_modulo_cores(
+	vrf_in_out: &VRFInOut,
+	// Configuration - `relay_vrf_modulo_samples`.
+	num_samples: u32,
+	// Configuration - `n_cores`.
+	max_cores: u32,
+) -> Vec<CoreIndex> {
+	if num_samples as usize > MAX_MODULO_SAMPLES {
+		gum::warn!(
+			target: LOG_TARGET,
+			n_cores = max_cores,
+			num_samples,
+			max_modulo_samples = MAX_MODULO_SAMPLES,
+			"`num_samples` is greater than `MAX_MODULO_SAMPLES`",
+		);
+	}
+
+	vrf_in_out
+		.make_bytes::<BigArray>(approval_types::v2::CORE_RANDOMNESS_CONTEXT)
```

Comment:
if we were to change the `MAX_MODULO_SAMPLES` later on, would the byte strings begin in the same way?

Reply:
That is my expectation, but I haven't checked.

Reply:
cc @burdges